deepdrivemd/aggregation/stream/aggregator.py

```diff
@@ -0,0 +1,209 @@
+    i = 0
+    for bp in bpfiles:
+        bp = bp.replace(".sst", "")
+        dir = os.path.dirname(bp)
+        task_md = os.path.basename(dir)
+        taskid_md = int(task_md.replace("task", ""))
+        adios_md = dir + "/adios.xml"
+
+        taskid_agg = cfg.task_idx
+
+        print(f"taskid_md = {taskid_md}, i = {i}, {i*bp_slice}, {(i+1)*bp_slice}")
+
+        if taskid_md // bp_slice == taskid_agg:
+            adios = adios2.ADIOS(adios_md, True)
+            io = adios.DeclareIO(task_md)
+            io.SetParameters({"ControlModule": "epoll"})
+            stream = io.Open(bp, adios2.Mode.Read)
+            connections[taskid_md] = (adios, io, stream)
+
+        i += 1
+
+    return connections
+
+
+def aggregate(cfg: StreamAggregation, connections, aggregator_stream):
+    """
+    Read adios streams from a subset of simulations handled by this aggregator and write them to adios file to be used
+    by machine learning and outlier search. If we do not need to save the data for postproduction, we can get rid of the aggregated
+    adios file and replace it by SST stream.
+    """
+
+    # infinite loop over simulation reporting steps
+    for iteration in itertools.count(0):
+        timer("aggregator_iteration", 1)
+        print("iteration = ", iteration)
+
+        q = queue.Queue()
+        for s in connections.keys():
+            q.put(s)
+
+        # Read data from each simulation and write it to the aggregated adios file
+        # If the data is not ready yet, go to the next simulation and revisit the current one later
+        while not q.empty():
+            sim_task_id = q.get()
+            adios, io, stream = connections[sim_task_id]
+
+            status = stream.BeginStep(adios2.StepMode.Read, 0.0)
+
+            if status == adios2.StepStatus.NotReady:
+                print(f"NotReady in simulation {sim_task_id}")
+                q.put(sim_task_id)
+                continue
+            if status == adios2.StepStatus.EndOfStream:
+                print(f"EndOfStream in simulation {sim_task_id}")
+                q.put(sim_task_id)
+                continue
+            if status == adios2.StepStatus.OtherError:
+                print(f"OtherError in simulation {sim_task_id}")
+                q.put(sim_task_id)
+                continue
+
+            step = np.zeros(1, dtype=np.int32)
+            varStep = io.InquireVariable("step")
+            stream.Get(varStep, step)
+
+            rmsd = np.zeros(1, dtype=np.float32)
+            varRMSD = io.InquireVariable("rmsd")
+            stream.Get(varRMSD, rmsd)
+
+            varCM = io.InquireVariable("contact_map")
+            shapeCM = varCM.Shape()
+            ndimCM = len(shapeCM)
+            start = [0] * ndimCM
+            count = shapeCM
+            varCM.SetSelection([start, count])
+            cm = np.zeros(shapeCM, dtype=np.int8)
+            stream.Get(varCM, cm)
+
+            varPositions = io.InquireVariable("positions")
```

Comment:
Could we make functions here like (maybe the name can change): 

```
def get_scalar(io, stream, name: str, dtype: type = np.float32) -> np.ndarray:
    var = io.InquireVariable(name)
    data = np.zeros(1, dtype=np.int32)
    stream.Get(var, data)
    return data


def get_array(io, stream, name: str, dtype: type = np.float32) -> np.ndarray:
    var = io.InquireVariable(name)
    shape = var.Shape()
    start = [0] * len(shape) # ndim
    var.SetSelection([start, shape])
    data = np.zeros(shape, dtype=dtype)
    stream.Get(var, data)
    return data
```

Then this could be simplified to: 
```
step = get_scalar(io, stream, "step", np.int32)
rmsd = get_scalar(io, stream, "rmsd", np.float32)
cm = get_array(io, stream, "contact_map", np.int8)
positions = get_array(io, stream, "positions", np.float32)
velocities = get_array(io, stream, "velocities", np.float32)
md5 = get_array(io, stream, "md5", np.int64)
```
This would be really helpful for generalizing this code.

Reply:
I do not think it is a good idea: things between BeginStep() and EndStep() are guaranteed to be finished only after EndStep() as far as I understand ADIOS. Having local var inside functions might be a problem: they might go out of scope before the data transfer happens. Passing var as a parameter to a function would only make things more complicated and ugly since all those var would have to be named differently anyway and InquireVariable() should be executed outside of get_...() functions.
numpy array would also probably have to be allocated outside of the function which requires knowledge of the shape...

Reply:
I'll ask ADIOS people if my understanding is correct.

Reply:
Oh the async bit makes it tricky then. In that case, can we drop these in a small class? That should preserve the variable state. We need a way to abstract this functionality which will be crucial for preprocessing arbitrary data types. Something like this might work:

```
class StreamVariable:
    def __init__(self, io, stream, name: str, dtype: type, scalar: bool):

        self.var = io.InquireVariable(name)
        
        if scalar:
            self._data = np.zeros(1, dtype=dtype)
            stream.Get(self.var, self._data)
        else:
            shape = self.var.Shape()
            start = [0] * len(shape) # ndim
            self.var.SetSelection([start, shape])
            self._data = np.zeros(shape, dtype=dtype)
            stream.Get(var, self._data)

    @property
    def data(self) -> np.ndarray:
        return self._data
       
       
       
 # Example:
step = StreamVariable(io, stream, "step", np.int32, True)
rmsd = StreamVariable(io, stream, "rmsd", np.float32, True)
cm = StreamVariable(io, stream, "contact_map", np.int8, False)
positions = StreamVariable(io, stream, "positions", np.float32, False)
velocities = StreamVariable(io, stream, "velocities", np.float32, False)
md5 = StreamVariable(io, stream, "md5", np.int64, False)

aggregator_stream.write("step", step.data)
```
What do you think?

Reply:
That has a better chance of working although I am still not sure if it might interfere with ADIOS peculiarities. I'll try to play with it and ask ADIOS people although they still did not answer my previous question.

Reply:
Thanks, Igor. If this abstraction works, we could use it in a few other areas of the code as well.

Reply:
I just update the class to include a write function which writes the self._data to another stream. This also saves on code duplication and should be reusable.

Reply:
So far all this is not working with streams: I am getting some ADIOS errors I have never seen before

*** Error in `/usr/workspace/cv_ddmd/conda1/powerai/bin/python': munmap_chunk(): invalid pointer: 0x00000001013a6000 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x91e04)[0x2000001d1e04]
/lib64/libc.so.6(cfree+0xd10)[0x2000001d9fa0]
/usr/workspace/cv_ddmd/conda1/powerai/lib/./libstdc++.so.6(_ZdlPv+0x18)[0x20000053706c]
/usr/workspace/cv_ddmd/software1/ADIOS2/lib/python3.6/site-packages/adios2/../../../../lib64/libadios2_core.so.2(_ZN6adios24core8VariableIfE4InfoD1Ev+0xbc)[0x20001098080c]
/usr/workspace/cv_ddmd/software1/ADIOS2/lib/python3.6/site-packages/adios2/../../../../lib64/libadios2_core.so.2(_ZN6adios24core6engine9SstReader11PerformGetsEv+0x107c)[0x200010d6fe4c]
/usr/workspace/cv_ddmd/software1/ADIOS2/lib/python3.6/site-packages/adios2/../../../../lib64/libadios2_core.so.2(_ZN6adios24core6engine9SstReader7EndStepEv+0x20c)[0x200010d7620c]
/

Reply:
Is this for the function or class implementation?

Reply:
My guess is that all the variables should be in the same scope between BeginStep() and EndStep().
So perhaps a different design might achieve this: all the variables are members of the same class; methods operate on those variables without passing them as arguments. 

But it would be even better to leave ADIOS as is.

Reply:
This is for the class implementation.

Reply:
This design at first glance is working. I'll commit it in a minute.

class ADIOS_RW_FULL_API:
    def __init__(self, connections, variables):
        self.connections = connections
        self.variables = variables
        
    def read_step(self, sim_task_id):
        adios, io, stream = self.connections[sim_task_id]

        status = stream.BeginStep(adios2.StepMode.Read, 0.0)

        if(not (status == adios2.StepStatus.OK)):
            return False

        for v in self.variables:
            vname = "var_" + v
            dname = "d_" + v
            dtype = self.variables[v][0]
            structure_type = self.variables[v][1]
            setattr(self, vname, io.InquireVariable(v))
            if(structure_type == 0): # scalar
                setattr(self, dname, np.zeros(1, dtype=dtype))
                stream.Get(getattr(self, vname), getattr(self, dname) )
            elif(structure_type == 1): # np.array
                shape = getattr(self, vname).Shape()
                start = [0] * len(shape) # ndim
                getattr(self, vname).SetSelection([start, shape])
                setattr(self, dname, np.zeros(shape, dtype=dtype))
                stream.Get(getattr(self, vname), getattr(self, dname) )
            else: # string
                setattr(self, dname, stream.Get(getattr(self, vname)))
        stream.EndStep()
        return True


    def write_step(self, wstream, variables, end_step = False):
        for v in variables:
            dname = "d_" + v
            dtype = variables[v][0]
            structure_type = variables[v][1]
            data = getattr(self, dname)
            end = False
            if(end_step and v == variables[-1]):
                end = True
                
            if(structure_type == 0): # scalar
                wstream.write(v, data, end_step = end)
            elif(structure_type == 1): # np.array
                wstream.write(v, data, list(data.shape), [0]*len(data.shape), list(data.shape), end_step = end)

Reply:
Thanks for getting this to work! I think having this logic abstracted will really help moving forward. Going to resolve this conversation.